src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1368,52 +1435,378 @@ private sealed class RemainingTestsForCase
+                        other: Test,
+                        whenTrueValues: whenTrueValues,
+                        whenFalseValues: whenFalseValues,
+                        syntax: test.Syntax,
+                        trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
+                        falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
+                        trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
+                        falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
+                        foundExplicitNullTest: ref foundExplicitNullTest);
+                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                }
+                public override BoundDagTest ComputeSelectedTest() => this.Test;
+                public override Tests RemoveEvaluation(BoundDagEvaluation e) => Test is BoundDagEvaluation e2 && e2 == e ? Tests.True.Instance : (Tests)this;
+                public override string Dump(Func<BoundDagTest, String> dump) => dump(this.Test);
+                public override bool Equals(object? obj)
+                {
+                    if (!(obj is One other))
+                        return false;
 
-                    case BoundKind.DagExplicitNullTest:
-                    case BoundKind.DagNonNullTest:
-                        return true;
+                    return sameTest(this.Test, other.Test);
 
-                    default:
-                        // For an evaluation, we defer to its .Equals
-                        return x.Equals(y);
+                    static bool sameTest(BoundDagTest x, BoundDagTest y)
+                    {
+                        if (x.Kind != y.Kind || x.Input != y.Input)
+                        {
+                            return false;
+                        }
+
+                        switch (x.Kind)
+                        {
+                            case BoundKind.DagTypeTest:
+                                return ((BoundDagTypeTest)x).Type.Equals(((BoundDagTypeTest)y).Type, TypeCompareKind.AllIgnoreOptions);
+
+                            case BoundKind.DagValueTest:
+                                return ((BoundDagValueTest)x).Value == ((BoundDagValueTest)y).Value;
+
+                            case BoundKind.DagExplicitNullTest:
+                            case BoundKind.DagNonNullTest:
+                                return true;
+
+                            default:
+                                // For an evaluation, we defer to its .Equals
+                                return x.Equals(y);
+                        }
+                    }
+                }
+                public override int GetHashCode()
```

Comment:
>GetHashCode [](start = 36, length = 11)

Is `Tests.GetHashCode()` used?

Reply:
Yes, indirectly.  See line 562:
```csharp
            var uniqueState = new Dictionary<DagState, DagState>(DagStateEquivalence.Instance);
```
Where `DagStateEquivalence` does
``` csharp
                return Hash.Combine(Hash.CombineValues(x.Cases), x.Cases.Length);
```
and in `StateForCase` we have
``` csharp
                return Hash.Combine(RemainingTests.GetHashCode(), Index);
```
where `RemainingTests` is of type `Tests`.


---
In reply to: [382859814](https://github.com/dotnet/roslyn/pull/41503#discussion_r382859814) [](ancestors = 382859814)